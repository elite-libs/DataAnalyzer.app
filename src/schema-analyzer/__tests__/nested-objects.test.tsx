import { schemaAnalyzer, extractNestedTypes } from '../';
import historicEvents from '../../../public/data/historic-events.json';
import pokemonCharmander from '../../../public/data/pokemon-charmander.json';

describe('#nested non-array object types', () => {
  // Uses data from http://history.muffinlabs.com/date/2/16
  it('can parse nested objects (non-array)', async () => {
    try {
      const result = await schemaAnalyzer('historicEvent', historicEvents);
      console.log(JSON.stringify(result, null, 2));
      expect(result).toMatchSnapshot();
      expect(Object.keys(result.fields)).toStrictEqual(['date', 'url', 'data']);
      expect(result.nestedTypes).not.toBeNull();
      expect(result.nestedTypes?.['historicEvent.data']).toBeDefined();
      expect(
        Object.keys(result.nestedTypes?.['historicEvent.data'].fields),
      ).toStrictEqual(['Events', 'Births', 'Deaths']);
    } catch (error) {
      console.error('ERROR:', error);
    }
  });
  // https://pokeapi.co/api/v2/pokemon/4
  // pokemonCharmander
});
