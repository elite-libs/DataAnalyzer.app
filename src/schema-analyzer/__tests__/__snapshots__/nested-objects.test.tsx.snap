// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#nested non-array object types can parse nested objects (non-array) 1`] = `
Object {
  "fields": Object {
    "data": Object {
      "enum": undefined,
      "identity": false,
      "nullCount": 0,
      "nullable": undefined,
      "types": Object {
        "$ref": Object {
          "count": 1,
          "typeAlias": "historicEvent.data",
        },
        "Object": Object {
          "count": 1,
        },
      },
      "unique": undefined,
      "uniqueCount": 0,
    },
    "date": Object {
      "enum": undefined,
      "identity": false,
      "nullCount": 0,
      "nullable": false,
      "types": Object {
        "String": Object {
          "count": 1,
          "length": Object {
            "max": 11,
            "mean": 11,
            "min": 11,
            "p25": 11,
            "p33": 11,
            "p50": 11,
            "p66": 11,
            "p75": 11,
            "p99": 11,
          },
        },
      },
      "unique": true,
      "uniqueCount": 1,
    },
    "url": Object {
      "enum": undefined,
      "identity": false,
      "nullCount": 0,
      "nullable": false,
      "types": Object {
        "String": Object {
          "count": 1,
          "length": Object {
            "max": 38,
            "mean": 38,
            "min": 38,
            "p25": 38,
            "p33": 38,
            "p50": 38,
            "p66": 38,
            "p75": 38,
            "p99": 38,
          },
        },
      },
      "unique": true,
      "uniqueCount": 1,
    },
  },
  "nestedTypes": Object {
    "historicEvent.data": Object {
      "fields": Object {
        "Births": Object {
          "enum": undefined,
          "identity": false,
          "nullCount": 0,
          "nullable": undefined,
          "types": Object {
            "$ref": Object {
              "count": 1,
              "typeAlias": "historicEvent.data.Births",
            },
            "Array": Object {
              "count": 1,
              "length": Object {
                "max": 178,
                "mean": 178,
                "min": 178,
                "p25": 178,
                "p33": 178,
                "p50": 178,
                "p66": 178,
                "p75": 178,
                "p99": 178,
              },
            },
          },
          "unique": undefined,
          "uniqueCount": 0,
        },
        "Deaths": Object {
          "enum": undefined,
          "identity": false,
          "nullCount": 0,
          "nullable": undefined,
          "types": Object {
            "$ref": Object {
              "count": 1,
              "typeAlias": "historicEvent.data.Deaths",
            },
            "Array": Object {
              "count": 1,
              "length": Object {
                "max": 76,
                "mean": 76,
                "min": 76,
                "p25": 76,
                "p33": 76,
                "p50": 76,
                "p66": 76,
                "p75": 76,
                "p99": 76,
              },
            },
          },
          "unique": undefined,
          "uniqueCount": 0,
        },
        "Events": Object {
          "enum": undefined,
          "identity": false,
          "nullCount": 0,
          "nullable": undefined,
          "types": Object {
            "$ref": Object {
              "count": 1,
              "typeAlias": "historicEvent.data.Events",
            },
            "Array": Object {
              "count": 1,
              "length": Object {
                "max": 40,
                "mean": 40,
                "min": 40,
                "p25": 40,
                "p33": 40,
                "p50": 40,
                "p66": 40,
                "p75": 40,
                "p99": 40,
              },
            },
          },
          "unique": undefined,
          "uniqueCount": 0,
        },
      },
      "nestedTypes": undefined,
      "schemaName": "historicEvent.data",
      "totalRows": 1,
    },
    "historicEvent.data.Births": Object {
      "fields": Object {
        "html": Object {
          "enum": undefined,
          "identity": false,
          "nullCount": 0,
          "nullable": false,
          "types": Object {
            "String": Object {
              "count": 178,
              "length": Object {
                "max": 462,
                "mean": 152.08426966292134,
                "min": 100,
                "p25": 127,
                "p33": 131,
                "p50": 140,
                "p66": 149,
                "p75": 156,
                "p99": 356,
              },
            },
          },
          "unique": true,
          "uniqueCount": 178,
        },
        "links": Object {
          "enum": undefined,
          "identity": false,
          "nullCount": 0,
          "nullable": undefined,
          "types": Object {
            "$ref": Object {
              "count": 178,
              "typeAlias": "historicEvent.data.Births.links",
            },
            "Array": Object {
              "count": 178,
              "length": Object {
                "max": 3,
                "mean": 1.0898876404494382,
                "min": 1,
                "p25": 1,
                "p33": 1,
                "p50": 1,
                "p66": 1,
                "p75": 1,
                "p99": 3,
              },
            },
          },
          "unique": undefined,
          "uniqueCount": 0,
        },
        "no_year_html": Object {
          "enum": undefined,
          "identity": false,
          "nullCount": 0,
          "nullable": false,
          "types": Object {
            "String": Object {
              "count": 178,
              "length": Object {
                "max": 455,
                "mean": 145.08426966292134,
                "min": 93,
                "p25": 120,
                "p33": 124,
                "p50": 133,
                "p66": 142,
                "p75": 149,
                "p99": 349,
              },
            },
          },
          "unique": true,
          "uniqueCount": 178,
        },
        "text": Object {
          "enum": undefined,
          "identity": false,
          "nullCount": 0,
          "nullable": false,
          "types": Object {
            "String": Object {
              "count": 178,
              "length": Object {
                "max": 118,
                "mean": 52.353932584269664,
                "min": 27,
                "p25": 41,
                "p33": 44,
                "p50": 51,
                "p66": 57,
                "p75": 61,
                "p99": 109,
              },
            },
          },
          "unique": true,
          "uniqueCount": 178,
        },
        "year": Object {
          "enum": undefined,
          "identity": false,
          "nullCount": 0,
          "nullable": false,
          "types": Object {
            "Number": Object {
              "count": 178,
              "value": Object {
                "max": 1995,
                "mean": 1898.938202247191,
                "min": 1222,
                "p25": 1891,
                "p33": 1916,
                "p50": 1944,
                "p66": 1961,
                "p75": 1974,
                "p99": 1995,
              },
            },
          },
          "unique": false,
          "uniqueCount": 116,
        },
      },
      "nestedTypes": undefined,
      "schemaName": "historicEvent.data.Births",
      "totalRows": 178,
    },
    "historicEvent.data.Births.links": Object {
      "fields": Object {
        "link": Object {
          "enum": undefined,
          "identity": false,
          "nullCount": 0,
          "nullable": false,
          "types": Object {
            "String": Object {
              "count": 194,
              "length": Object {
                "max": 85,
                "mean": 43.75773195876289,
                "min": 32,
                "p25": 39,
                "p33": 40,
                "p50": 41,
                "p66": 44,
                "p75": 47,
                "p99": 71,
              },
            },
          },
          "unique": true,
          "uniqueCount": 194,
        },
        "title": Object {
          "enum": undefined,
          "identity": false,
          "nullCount": 0,
          "nullable": false,
          "types": Object {
            "String": Object {
              "count": 194,
              "length": Object {
                "max": 50,
                "mean": 16.298969072164947,
                "min": 5,
                "p25": 12,
                "p33": 13,
                "p50": 14,
                "p66": 16,
                "p75": 18,
                "p99": 44,
              },
            },
          },
          "unique": true,
          "uniqueCount": 194,
        },
      },
      "nestedTypes": undefined,
      "schemaName": "historicEvent.data.Births.links",
      "totalRows": 194,
    },
    "historicEvent.data.Deaths": Object {
      "fields": Object {
        "html": Object {
          "enum": undefined,
          "identity": false,
          "nullCount": 0,
          "nullable": false,
          "types": Object {
            "String": Object {
              "count": 76,
              "length": Object {
                "max": 494,
                "mean": 170.5,
                "min": 111,
                "p25": 134,
                "p33": 139,
                "p50": 148,
                "p66": 154,
                "p75": 161,
                "p99": 494,
              },
            },
          },
          "unique": true,
          "uniqueCount": 76,
        },
        "links": Object {
          "enum": undefined,
          "identity": false,
          "nullCount": 0,
          "nullable": undefined,
          "types": Object {
            "$ref": Object {
              "count": 76,
              "typeAlias": "historicEvent.data.Deaths.links",
            },
            "Array": Object {
              "count": 76,
              "length": Object {
                "max": 3,
                "mean": 1.1973684210526316,
                "min": 1,
                "p25": 1,
                "p33": 1,
                "p50": 1,
                "p66": 1,
                "p75": 1,
                "p99": 3,
              },
            },
          },
          "unique": undefined,
          "uniqueCount": 0,
        },
        "no_year_html": Object {
          "enum": undefined,
          "identity": false,
          "nullCount": 0,
          "nullable": false,
          "types": Object {
            "String": Object {
              "count": 76,
              "length": Object {
                "max": 487,
                "mean": 163.52631578947367,
                "min": 104,
                "p25": 127,
                "p33": 132,
                "p50": 141,
                "p66": 147,
                "p75": 154,
                "p99": 487,
              },
            },
          },
          "unique": true,
          "uniqueCount": 76,
        },
        "text": Object {
          "enum": undefined,
          "identity": false,
          "nullCount": 0,
          "nullable": false,
          "types": Object {
            "String": Object {
              "count": 76,
              "length": Object {
                "max": 148,
                "mean": 59.38157894736842,
                "min": 32,
                "p25": 47,
                "p33": 49,
                "p50": 57,
                "p66": 64,
                "p75": 67,
                "p99": 148,
              },
            },
          },
          "unique": true,
          "uniqueCount": 76,
        },
        "year": Object {
          "enum": undefined,
          "identity": false,
          "nullCount": 0,
          "nullable": false,
          "types": Object {
            "Number": Object {
              "count": 76,
              "value": Object {
                "max": 2019,
                "mean": 1862.6184210526317,
                "min": 549,
                "p25": 1907,
                "p33": 1932,
                "p50": 1992,
                "p66": 2001,
                "p75": 2009,
                "p99": 2019,
              },
            },
          },
          "unique": false,
          "uniqueCount": 54,
        },
      },
      "nestedTypes": undefined,
      "schemaName": "historicEvent.data.Deaths",
      "totalRows": 76,
    },
    "historicEvent.data.Deaths.links": Object {
      "fields": Object {
        "link": Object {
          "enum": undefined,
          "identity": false,
          "nullCount": 0,
          "nullable": false,
          "types": Object {
            "String": Object {
              "count": 91,
              "length": Object {
                "max": 85,
                "mean": 44.86813186813187,
                "min": 36,
                "p25": 39,
                "p33": 40,
                "p50": 42,
                "p66": 45,
                "p75": 48,
                "p99": 85,
              },
            },
          },
          "unique": true,
          "uniqueCount": 91,
        },
        "title": Object {
          "enum": undefined,
          "identity": false,
          "nullCount": 0,
          "nullable": false,
          "types": Object {
            "String": Object {
              "count": 91,
              "length": Object {
                "max": 53,
                "mean": 17.01098901098901,
                "min": 9,
                "p25": 12,
                "p33": 13,
                "p50": 14,
                "p66": 17,
                "p75": 19,
                "p99": 53,
              },
            },
          },
          "unique": true,
          "uniqueCount": 91,
        },
      },
      "nestedTypes": undefined,
      "schemaName": "historicEvent.data.Deaths.links",
      "totalRows": 91,
    },
    "historicEvent.data.Events": Object {
      "fields": Object {
        "html": Object {
          "enum": undefined,
          "identity": false,
          "nullCount": 0,
          "nullable": false,
          "types": Object {
            "String": Object {
              "count": 40,
              "length": Object {
                "max": 730,
                "mean": 342.85,
                "min": 97,
                "p25": 265,
                "p33": 275,
                "p50": 344,
                "p66": 365,
                "p75": 414,
                "p99": 730,
              },
            },
          },
          "unique": true,
          "uniqueCount": 40,
        },
        "links": Object {
          "enum": undefined,
          "identity": false,
          "nullCount": 0,
          "nullable": undefined,
          "types": Object {
            "$ref": Object {
              "count": 40,
              "typeAlias": "historicEvent.data.Events.links",
            },
            "Array": Object {
              "count": 40,
              "length": Object {
                "max": 5,
                "mean": 2.75,
                "min": 1,
                "p25": 2,
                "p33": 2,
                "p50": 3,
                "p66": 3,
                "p75": 3,
                "p99": 5,
              },
            },
          },
          "unique": undefined,
          "uniqueCount": 0,
        },
        "no_year_html": Object {
          "enum": undefined,
          "identity": false,
          "nullCount": 0,
          "nullable": false,
          "types": Object {
            "String": Object {
              "count": 40,
              "length": Object {
                "max": 723,
                "mean": 335.85,
                "min": 90,
                "p25": 258,
                "p33": 268,
                "p50": 337,
                "p66": 358,
                "p75": 407,
                "p99": 723,
              },
            },
          },
          "unique": true,
          "uniqueCount": 40,
        },
        "text": Object {
          "enum": undefined,
          "identity": false,
          "nullCount": 0,
          "nullable": false,
          "types": Object {
            "String": Object {
              "count": 40,
              "length": Object {
                "max": 190,
                "mean": 92.475,
                "min": 21,
                "p25": 72,
                "p33": 75,
                "p50": 90,
                "p66": 98,
                "p75": 107,
                "p99": 190,
              },
            },
          },
          "unique": true,
          "uniqueCount": 40,
        },
        "year": Object {
          "enum": undefined,
          "identity": false,
          "nullCount": 0,
          "nullable": false,
          "types": Object {
            "Number": Object {
              "count": 40,
              "value": Object {
                "max": 2013,
                "mean": 1884.475,
                "min": 1249,
                "p25": 1881,
                "p33": 1923,
                "p50": 1943,
                "p66": 1962,
                "p75": 1985,
                "p99": 2013,
              },
            },
          },
          "unique": false,
          "uniqueCount": 37,
        },
      },
      "nestedTypes": undefined,
      "schemaName": "historicEvent.data.Events",
      "totalRows": 40,
    },
    "historicEvent.data.Events.links": Object {
      "fields": Object {
        "link": Object {
          "enum": undefined,
          "identity": false,
          "nullCount": 0,
          "nullable": false,
          "types": Object {
            "String": Object {
              "count": 110,
              "length": Object {
                "max": 79,
                "mean": 45.03636363636364,
                "min": 31,
                "p25": 37,
                "p33": 39,
                "p50": 44,
                "p66": 48,
                "p75": 51,
                "p99": 75,
              },
            },
          },
          "unique": false,
          "uniqueCount": 109,
        },
        "title": Object {
          "enum": undefined,
          "identity": false,
          "nullCount": 0,
          "nullable": false,
          "types": Object {
            "String": Object {
              "count": 110,
              "length": Object {
                "max": 41,
                "mean": 17.327272727272728,
                "min": 4,
                "p25": 10,
                "p33": 12,
                "p50": 17,
                "p66": 20,
                "p75": 23,
                "p99": 40,
              },
            },
          },
          "unique": false,
          "uniqueCount": 109,
        },
      },
      "nestedTypes": undefined,
      "schemaName": "historicEvent.data.Events.links",
      "totalRows": 110,
    },
  },
  "schemaName": "historicEvent",
  "totalRows": 1,
}
`;
